example/particles/particles.cpp

```diff
@@ -0,0 +1,362 @@
+//========================================================================================
+// (C) (or copyright) 2020. Triad National Security, LLC. All rights reserved.
+//
+// This program was produced under U.S. Government contract 89233218CNA000001 for Los
+// Alamos National Laboratory (LANL), which is operated by Triad National Security, LLC
+// for the U.S. Department of Energy/National Nuclear Security Administration. All rights
+// in the program are reserved by Triad National Security, LLC, and the U.S. Department
+// of Energy/National Nuclear Security Administration. The Government is granted for
+// itself and others acting on its behalf a nonexclusive, paid-up, irrevocable worldwide
+// license in this material to reproduce, prepare derivative works, distribute copies to
+// the public, perform publicly and display publicly, and to permit others to do so.
+//========================================================================================
+
+#include "particles.hpp"
+
+#include <algorithm>
+#include <limits>
+#include <memory>
+#include <string>
+#include <utility>
+#include <vector>
+
+// *************************************************//
+// redefine some internal parthenon functions      *//
+// *************************************************//
+namespace particles_example {
+
+Packages_t ProcessPackages(std::unique_ptr<ParameterInput> &pin) {
+  Packages_t packages;
+  packages["particles_package"] = particles_example::Particles::Initialize(pin.get());
+  return packages;
+}
+
+void ProblemGenerator(MeshBlock *pmb, ParameterInput *pin) {
+  // Don't do anything for now
+}
+
+// *************************************************//
+// define the "physics" package particles_package, *//
+// which includes defining various functions that  *//
+// control how parthenon functions and any tasks   *//
+// needed to implement the "physics"               *//
+// *************************************************//
+
+namespace Particles {
+
+std::shared_ptr<StateDescriptor> Initialize(ParameterInput *pin) {
+  auto pkg = std::make_shared<StateDescriptor>("particles_package");
+
+  int num_particles = pin->GetOrAddInteger("Particles", "num_particles", 100);
+  pkg->AddParam<>("num_particles", num_particles);
+  Real particle_speed = pin->GetOrAddReal("Particles", "particle_speed", 1.0);
+  pkg->AddParam<>("particle_speed", particle_speed);
+  Real const_dt = pin->GetOrAddReal("Particles", "const_dt", 1.0);
+  pkg->AddParam<>("const_dt", const_dt);
+
+  // Initialize random number generator pool
+  int rng_seed = pin->GetInteger("Particles", "rng_seed");
+  pkg->AddParam<>("rng_seed", rng_seed);
+  RNGPool rng_pool(rng_seed);
+  pkg->AddParam<>("rng_pool", rng_pool);
+
+  std::string swarm_name = "my particles";
+  Metadata swarm_metadata;
+  pkg->AddSwarm(swarm_name, swarm_metadata);
```

Comment:
Why is `AddSwarm` separate from `AddSwarmValue` is there a need to have multiple swarms?

Reply:
`AddSwarm` adds an entire swarm of particles, and `AddSwarmValue` adds a new variable (a new pool of memory to have one scalar for each particle) to the swarm named in the `swarm_name` argument. 

Multiple swarms would be needed if you had multiple different species of particles.

Reply:
Oh I see. You're imagining a swarm per species, as opposed to species being an integer flag? Yeah that makes sense.